File: backend/app/Models/Booking.php
<?php

namespace App\Models;

use App\Enums\BookingStatus;
use App\Enums\BookingType;
use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\HasOne;
use Illuminate\Support\Str;

class Booking extends Model
{
    use HasFactory, HasUuids;

    protected $fillable = [
        'reference', 'car_id', 'customer_id', 'pickup_at', 'return_at',
        'daily_rate', 'total_days', 'subtotal', 'addons_total', 'service_fee',
        'tax', 'discount', 'total_amount', 'addons', 'pickup_address',
        'booking_type', 'status', 'cancellation_reason', 'confirmed_at',
        'cancelled_at', 'host_response_deadline',
    ];

    protected function casts(): array
    {
        return [
            'status' => BookingStatus::class,
            'booking_type' => BookingType::class,
            'pickup_at' => 'datetime',
            'return_at' => 'datetime',
            'confirmed_at' => 'datetime',
            'cancelled_at' => 'datetime',
            'host_response_deadline' => 'datetime',
            'daily_rate' => 'decimal:2',
            'subtotal' => 'decimal:2',
            'addons_total' => 'decimal:2',
            'service_fee' => 'decimal:2',
            'tax' => 'decimal:2',
            'discount' => 'decimal:2',
            'total_amount' => 'decimal:2',
            'addons' => 'array',
        ];
    }

    protected static function booted(): void
    {
        static::creating(function (Booking $booking) {
            if (empty($booking->reference)) {
                $booking->reference = 'DRV-' . strtoupper(Str::random(8));
            }
        });
    }

    public function car(): BelongsTo { return $this->belongsTo(Car::class); }
    public function customer(): BelongsTo { return $this->belongsTo(User::class, 'customer_id'); }
    public function payment(): HasOne { return $this->hasOne(Payment::class); }
    public function trip(): HasOne { return $this->hasOne(Trip::class); }
    public function reviews(): HasMany { return $this->hasMany(Review::class); }
    public function disputes(): HasMany { return $this->hasMany(Dispute::class); }

    public function isPending(): bool { return $this->status === BookingStatus::Pending; }
    public function isConfirmed(): bool { return $this->status === BookingStatus::Confirmed; }
    public function isActive(): bool { return $this->status === BookingStatus::Active; }
    public function isCompleted(): bool { return $this->status === BookingStatus::Completed; }

    public function isCancellable(): bool
    {
        return in_array($this->status, [BookingStatus::Pending, BookingStatus::Confirmed]);
    }
}

File: frontend/lib/core/models/booking.dart
import '../utils/json_utils.dart';
import 'car.dart';
import 'review.dart';
import 'trip.dart';
import 'user.dart';

/// Price breakdown for a booking (matches `BookingResource.pricing` and the
/// standalone `POST /customer/bookings/pricing` response).
class BookingPricing {
  final double dailyRate;
  final int totalDays;
  final double subtotal;
  final double addonsTotal;
  final double serviceFee;
  final double tax;
  final double discount;
  final double totalAmount;

  const BookingPricing({
    this.dailyRate = 0,
    this.totalDays = 0,
    this.subtotal = 0,
    this.addonsTotal = 0,
    this.serviceFee = 0,
    this.tax = 0,
    this.discount = 0,
    this.totalAmount = 0,
  });

  factory BookingPricing.fromJson(Map<String, dynamic> json) => BookingPricing(
        dailyRate: asDouble(json['daily_rate']),
        totalDays: asInt(json['total_days']),
        subtotal: asDouble(json['subtotal']),
        addonsTotal: asDouble(json['addons_total']),
        serviceFee: asDouble(json['service_fee']),
        tax: asDouble(json['tax']),
        discount: asDouble(json['discount']),
        // Tolerate alternative keys from the pricing-preview endpoint.
        totalAmount: asDouble(
          json['total_amount'] ?? json['total'] ?? json['grand_total'],
        ),
      );
}

/// A rental booking (matches `BookingResource`).
class Booking {
  final String id;
  final String reference;
  final String carId;
  final String customerId;
  final String
      status; // pending | confirmed | active | completed | cancelled | declined
  final String bookingType; // instant | request
  final DateTime? pickupAt;
  final DateTime? returnAt;
  final String? pickupAddress;
  final BookingPricing pricing;
  final List<dynamic>? addons;
  final String? cancellationReason;
  final DateTime? confirmedAt;
  final DateTime? cancelledAt;
  final DateTime? hostResponseDeadline;
  final Car? car;
  final UserSummary? customer;
  final Map<String, dynamic>? payment;
  final Trip? trip;
  final List<Review> reviews;
  final DateTime? createdAt;

  const Booking({
    required this.id,
    required this.reference,
    required this.carId,
    required this.customerId,
    required this.status,
    this.bookingType = 'instant',
    this.pickupAt,
    this.returnAt,
    this.pickupAddress,
    this.pricing = const BookingPricing(),
    this.addons,
    this.cancellationReason,
    this.confirmedAt,
    this.cancelledAt,
    this.hostResponseDeadline,
    this.car,
    this.customer,
    this.payment,
    this.trip,
    this.reviews = const [],
    this.createdAt,
  });

  bool get isPending => status == 'pending';
  bool get isConfirmed => status == 'confirmed';
  bool get isActive => status == 'active';
  bool get isCompleted => status == 'completed';
  bool get isCancelled => status == 'cancelled' || status == 'declined';
  bool get canCancel => isPending || isConfirmed;
  bool get canReview => isCompleted && reviews.isEmpty;
  bool get isPaid => payment?['status'] == 'succeeded';

  int get totalDays {
    if (pricing.totalDays > 0) return pricing.totalDays;
    if (pickupAt != null && returnAt != null) {
      return returnAt!.difference(pickupAt!).inDays.clamp(1, 365);
    }
    return 0;
  }

  factory Booking.fromJson(Map<String, dynamic> json) => Booking(
        id: asString(json['id']),
        reference: asString(json['reference']),
        carId: asString(json['car_id']),
        customerId: asString(json['customer_id']),
        status: asString(json['status'], fallback: 'pending'),
        bookingType: asString(json['booking_type'], fallback: 'instant'),
        pickupAt: asDateTime(json['pickup_at']),
        returnAt: asDateTime(json['return_at']),
        pickupAddress: asStringOrNull(json['pickup_address']),
        pricing: json['pricing'] is Map
            ? BookingPricing.fromJson(
                Map<String, dynamic>.from(json['pricing']))
            : const BookingPricing(),
        addons: json['addons'] is List ? json['addons'] as List : null,
        cancellationReason: asStringOrNull(json['cancellation_reason']),
        confirmedAt: asDateTime(json['confirmed_at']),
        cancelledAt: asDateTime(json['cancelled_at']),
        hostResponseDeadline: asDateTime(json['host_response_deadline']),
        car: json['car'] is Map
            ? Car.fromJson(Map<String, dynamic>.from(json['car']))
            : null,
        customer: json['customer'] is Map
            ? UserSummary.fromJson(Map<String, dynamic>.from(json['customer']))
            : null,
        payment: asMap(json['payment']),
        trip: json['trip'] is Map
            ? Trip.fromJson(Map<String, dynamic>.from(json['trip']))
            : null,
        reviews: asList(json['reviews'], Review.fromJson),
        createdAt: asDateTime(json['created_at']),
      );
}
